crates/ruff_db/src/file_system/memory.rs

```diff
@@ -1,136 +1,473 @@
+                file_type: FileType::Directory,
+            },
+        };
+
+        Ok(metadata)
     }
 
     fn read(&self, path: &FileSystemPath) -> Result<String> {
-        let entry = self
-            .inner
-            .files
-            .get(&path.to_path_buf())
-            .ok_or_else(|| std::io::Error::new(ErrorKind::NotFound, "File not found"))?;
+        let by_path = self.inner.by_path.read().unwrap();
+        let normalized = normalize_path(path, &self.inner.cwd);
 
-        let value = entry.value();
+        let entry = by_path.get(&normalized).ok_or_else(not_found)?;
 
-        Ok(value.content.clone())
+        match entry {
+            Entry::File(file) => Ok(file.content.clone()),
+            Entry::Directory(_) => Err(is_a_directory()),
+        }
     }
 
     fn exists(&self, path: &FileSystemPath) -> bool {
-        self.inner.files.contains_key(&path.to_path_buf())
+        let by_path = self.inner.by_path.read().unwrap();
+        let normalized = normalize_path(path, &self.inner.cwd);
+
+        by_path.contains_key(&normalized)
+    }
+}
+
+impl Default for MemoryFileSystem {
+    fn default() -> Self {
+        MemoryFileSystem::new()
     }
 }
 
 impl std::fmt::Debug for MemoryFileSystem {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        let mut map = f.debug_map();
+        let paths = self.inner.by_path.read().unwrap();
 
-        for entry in self.inner.files.iter() {
-            map.entry(entry.key(), entry.value());
-        }
-        map.finish()
+        f.debug_map().entries(paths.iter()).finish()
     }
 }
 
-#[derive(Default)]
 struct MemoryFileSystemInner {
-    files: FxDashMap<FileSystemPathBuf, FileData>,
+    by_path: RwLock<FxHashMap<Utf8PathBuf, Entry>>,
+    cwd: Utf8PathBuf,
 }
 
 #[derive(Debug)]
-struct FileData {
+enum Entry {
+    File(File),
+    Directory(Directory),
+}
+
+impl Entry {
+    const fn is_file(&self) -> bool {
+        matches!(self, Entry::File(_))
+    }
+}
+
+#[derive(Debug)]
+struct File {
     content: String,
     last_modified: FileTime,
-    permission: u32,
+}
+
+#[derive(Debug)]
+struct Directory {
+    last_modified: FileTime,
+}
+
+fn not_found() -> std::io::Error {
+    std::io::Error::new(std::io::ErrorKind::NotFound, "No such file or directory")
+}
+
+fn is_a_directory() -> std::io::Error {
+    // Note: Rust returns `ErrorKind::IsADirectory` for this error but this is a nightly only variant :(.
+    //   So we have to use other for now.
+    std::io::Error::new(std::io::ErrorKind::Other, "Is a directory")
+}
+
+fn not_a_directory() -> std::io::Error {
+    // Note: Rust returns `ErrorKind::NotADirectory` for this error but this is a nightly only variant :(.
+    //   So we have to use `Other` for now.
+    std::io::Error::new(std::io::ErrorKind::Other, "Not a directory")
+}
+
+/// Normalizes the path by removing `.` and `..` components and transform the path into an absolute path.
+///
```

Comment:
I am very surprised that we have to implement such a thing ourselves instead of just using a battle-tested library method for this.

Reply:
I think it exists, it's just not stabalized. @BurntSushi knows more ;)